hazelcast/src/main/java/com/hazelcast/map/impl/SizeEstimator.java

```diff
@@ -16,18 +16,101 @@
 
 package com.hazelcast.map.impl;
 
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.nio.UnsafeHelper;
+import sun.misc.Unsafe;
+
+import javax.management.MBeanServer;
+import javax.management.ObjectName;
+import javax.management.openmbean.CompositeDataSupport;
+import java.lang.management.ManagementFactory;
+
 /**
  * Size estimator general contract.
  *
  * @param <T> the type of object which's size going to be estimated.
  */
-public interface SizeEstimator<T> {
+public abstract class SizeEstimator<T> {
+
+    public static final boolean HAS_COMPRESSED_OOPS = hasCompressedOops();
+
+    @SuppressWarnings("CheckStyle")
+    public static final int OBJECT_HEADER_IN_BYTES = HAS_COMPRESSED_OOPS ? 12 : 16;
```

Comment:
In fact, there is no public way to detect compressed-oops even though via `Unsafe`.
The best way is through Hotspot SA agent but it is hard to use in embedded mode.
Another hacky way is Andrei Pangin's `Helfy`: https://github.com/apangin/helfy/
As far as I have experienced, it can be used to access JVM internal symbols so you can get some JVM details without Hotspot SA agent